### PR TITLE
APEXMALHAR-2459 1)Refactor the existing Kafka Input Operator. 2)Added the support of KafkaInputOperator using 0.10 consumer API

### PR DESCRIPTION
**Pull request by chaithu14**

### CODE REVIEW

**Comment by chaithu14** [comment]
@siyuanh @tushargosavi @tweise Please review and merge.

**Comment by tushargosavi** [comment]
@chaithu14 
Can you fix following error while compilation.

[ERROR] Failed to execute goal on project malhar-examples-kafka: Could not resolve dependencies for project org.apache.apex:malhar-examples-kafka:jar:3.8.0-SNAPSHOT: Failure to find org.apache.apex:malhar-kafka:jar:3.8.0-SNAPSHOT in http://repository.apache.org/snapshots was cached in the local repository, resolution will not be reattempted until the update interval of apache.snapshots has elapsed or updates are forced -> [Help 1]

**Comment by tweise** [comment]
Probably that is an issue with backward compatibility that should be not be affected unless there was a reason and discussion to the contrary.

**Comment by chaithu14** [comment]
@tushargosavi : Will hit this issue, when re-base the PR. I will fix it and will update the PR.

@tweise: I renamed the artifactId from malhar-kafka to malhar-kafka09. But, the kafka operator is still under "org.apache.apex.malhar.kafka" package, which i didn't change. Will this change comes under backward compatibility? If yes, then I will make the changes in the following way:
1) Copy the existing malhar-kafka to malhar-kafka09
2) Mark the malhar-kafka as deprecated.

**Comment by tweise** [comment]
I don't see any need to rename the existing module and certainly not to copy/duplicate it.

**Comment by chaithu14** [comment]
@tweise To make consistent between the two versions(Kafka-0.9, Kafka-0.10), i renamed the existing module.

**Comment by tweise** [comment]
I don't think that is appropriate as it breaks existing usage of the module.

**Comment by chaithu14** [comment]
@tweise  Thanks for you comments. I realized after reading the document https://apex.apache.org/docs/apex/compatibility/, the above changes comes under backward compatible.
I will retain the existing module and will update the PR.

**Comment by chaithu14** [comment]
I looked into other projects for fixing similar issue without duplicating the code. 
Below is the summary:
1) Renamed the existing operators and module. For example, AbstractKafkaInputOperator renamed to AbstractKafkaInput09Operator, .... malhar-kafka to malhar-kafka09, etc.
2) Create a malhar-kafka module and mark them as deprecated. In this module, added the existing operators which  extends from *09Operator. For example, AbstractKafkaInputOperator extends from AbstractKafkaInput09Operator and mark them as deprecated. This way, all the operators in this module marked them as deprecated and are compatible.

@tweise @tushargosavi Please share your thoughts on this.

File: kafka/kafka-common/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java
Comment (tweise) [inline kafka/kafka-common/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java]: Retain the original class name. I see no need to change it.

File: kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java
Comment (tweise) [inline kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java]: Include 010 here instead.

File: kafka/kafka09/pom.xml
Comment (tweise) [inline kafka/kafka09/pom.xml]: use latest patch

**Comment by tweise** [comment]
@chaithu14 can you look at:

https://issues.apache.org/jira/browse/APEXMALHAR-2162
https://issues.apache.org/jira/browse/APEXMALHAR-2446
http://apache-apex-developers-list.78491.x6.nabble.com/EmbeddedKafka-Server-for-testing-kafka-related-code-td10672.html

**Comment by tweise** [comment]
This will also require update to the documentation (including embedded links)

File: kafka/kafka-common/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java
Comment (chaithu14) [reply kafka/kafka-common/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumerWrapper.java]: AbstractKafkaInputOperator spitted into 2 level hierarchy. AbstractBaseKafkaInputOperator consists of common design which is under kafka-common module and contains the createConsumer(...) abstract method. AbstractKafkaInputOperator extends from AbstractBaseKafkaInputOperator which is under malhar-kafka* modules.

File: kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java
Comment (chaithu14) [reply kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java]: Done.

File: kafka/kafka09/pom.xml
Comment (chaithu14) [reply kafka/kafka09/pom.xml]: Done.

**Comment by chaithu14** [comment]
@tweise Sure. I look into it and will open a separate PR for the above JIRA's.

**Comment by chaithu14** [comment]
@tweise @tushargosavi Updated the PR. Please review.

File: kafka-connectors/kafka-common/src/main/java/org/apache/apex/malhar/kafka/AbstractBaseKafkaInputOperator.java
Comment (tweise) [inline kafka-connectors/kafka-common/src/main/java/org/apache/apex/malhar/kafka/AbstractBaseKafkaInputOperator.java]: Please undo this rename.

Comment (chaithu14) [reply kafka-connectors/kafka-common/src/main/java/org/apache/apex/malhar/kafka/AbstractBaseKafkaInputOperator.java]: This change is needed.
Changed the class hierarchy from AbstractKafkaInputOperator <-- KafkaSinglePortInputOperator to
AbstractBaseKafkaInputOperator <-- AbstractKafkaInputOperator(malhar-kafka) <-- KafkaSinglePortInputOperator(malhar-kafka).
Please note, the functionality of AbstractKafkaInputOperator is same in both the hierarchies.
AbstractBaseKafkaInputOperator which is under malhar-common package contains createConsumer(...) abstract method.
AbstractKafkaInputOperator which is in malhar-kafka(supports 0.9 version) package creates the type of consumer based on the kafka version i.e, 0.9 consumer wrapper.

File: kafka-connectors/kafka010/src/test/java/org/apache/apex/malhar/kafka/KafkaConsumerPropertiesTest.java
Comment (tushargosavi) [inline kafka-connectors/kafka010/src/test/java/org/apache/apex/malhar/kafka/KafkaConsumerPropertiesTest.java]: This test seems to be repeat. Can it be moved to common package?

File: kafka-connectors/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafka010InputOperator.java
Comment (tushargosavi) [inline kafka-connectors/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafka010InputOperator.java]: I feel, it should be easy for user to move from 0.9 to 0.10 version by just changing the application pom.xml , If the class name is different then source code change is also needed , @tweise what do you think about following class structures.

- AbstractKafkaInputOperator  (In malhar-kafka010 artifact for 0.10 kafka)
- AbstractKafkaInputOperator (In malhar-kafka artifact for 0.9 kafka)
- AbstractBaseKafkaInputOperator (in malhar-kafka-common artifact)

**Review by tushargosavi — COMMENTED** [review]
@chaithu14 How was testing done? Have your verified read/write from kafka 0.9 and 0.10 versions? Also how backward compatibility was tested?

**Comment by tushargosavi** [comment]
@chaithu14 Documentation update is needed with this change.

Also there are no output operators in malhar-kafka10 , If I want to use output operators with kafka 10, what is the procedure? will adding dependency on malhar-kafka and excluding kaka-client (0.9) version will work?

File: kafka-connectors/kafka010/src/test/java/org/apache/apex/malhar/kafka/KafkaConsumerPropertiesTest.java
Comment (chaithu14) [reply kafka-connectors/kafka010/src/test/java/org/apache/apex/malhar/kafka/KafkaConsumerPropertiesTest.java]: Yes. You are right. Common test code can be moved to malhar-common package.

File: kafka-connectors/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafka010InputOperator.java
Comment (chaithu14) [reply kafka-connectors/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafka010InputOperator.java]: +1 for @tushargosavi point. 
@tweise Please share your opinion on this. If you are okay, i will rename to AbstractKafkaInputOperator.

**Comment by chaithu14** [comment]
@tushargosavi Below are my comments:

Testing the operator
===============
I tested kafka2hdfs example application which is under examples package. This application is referring to the 0.9 version of kafka client. 
I tested the same example by changing the artifactId from malhar-kafka to malhar-kafka010. It worked fine.

Documentation Update
=================
I think once the hierarchy and class names are fine, then it's good to start the documentation update. So, I will work on documentation update once this PR is merged. If you are okay with this, will create the JIRA for the same.

Output operators in Kafka-10
=====================
This JIRA is about the KafkaInputOperator. In this PR, there is no output operator for Kafka 0.10.  I will create JIRA for this task.

File: kafka-connectors/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafka010InputOperator.java
Comment (chaithu14) [reply kafka-connectors/kafka010/src/main/java/org/apache/apex/malhar/kafka/AbstractKafka010InputOperator.java]: @tweise Please share your opinion on the above class name.

**Comment by chaithu14** [comment]
@tushargosavi @tweise Created JIRA's for the above two tasks(APEXMALHAR-2496, APEXMALHAR-2497) and updated the PR. Please review and merge.

File: kafka-connectors/kafka-common/pom.xml
Comment (tweise) [inline kafka-connectors/kafka-common/pom.xml]: I see no reason to change the top level directory name.

File: kafka-connectors/kafka-common/pom.xml
Comment (tweise) [inline kafka-connectors/kafka-common/pom.xml]: Should probably use latest patch.

File: kafka-connectors/kafka09/pom.xml
Comment (tweise) [inline kafka-connectors/kafka09/pom.xml]: 0.9.0.0 -> 0.9

File: kafka-connectors/kafka-common/src/main/java/org/apache/apex/malhar/kafka/AbstractBaseKafkaInputOperator.java
Comment (tweise) [reply kafka-connectors/kafka-common/src/main/java/org/apache/apex/malhar/kafka/AbstractBaseKafkaInputOperator.java]: I don't see any need for the rename. I also don't see the need for derived abstract classes when the only difference is the creation of the consumer (you can add that to KafkaSinglePortInputOperator). Also, existing abstract class was evolving, the derived classes you created are not. Any new classes should be evolving.

**Comment by chaithu14** [comment]
@tweise Incorporated review comments. Please review the changes.

**Comment by chaithu14** [comment]
@tweise @tushargosavi Please review. If you are okay with the changes, I will squash the commits and then update the PR.

File: kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumer010.java
Comment (tweise) [inline kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumer010.java]: 0.10.x - check other occurrences

File: kafka/kafka010/src/test/resources/log4j.properties
Comment (tweise) [inline kafka/kafka010/src/test/resources/log4j.properties]: why is this commented out

File: kafka/kafka09/pom.xml
Comment (tweise) [inline kafka/kafka09/pom.xml]: do we really need a repetition of this and other boilerplate code in every module vs. having it defined in the parent

**Review by tweise — CHANGES_REQUESTED** [review]

File: kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumer010.java
Comment (chaithu14) [reply kafka/kafka010/src/main/java/org/apache/apex/malhar/kafka/KafkaConsumer010.java]: I verified and updated the version.

File: kafka/kafka09/pom.xml
Comment (chaithu14) [reply kafka/kafka09/pom.xml]: Yes. It's not needed in every module. I moved it to parent pom.

File: kafka/kafka010/src/test/resources/log4j.properties
Comment (chaithu14) [reply kafka/kafka010/src/test/resources/log4j.properties]: This is commented out even in the master branch.

Comment (tweise) [reply kafka/kafka010/src/test/resources/log4j.properties]: Unless you are aware why let's make it consistent with other modules.

Comment (chaithu14) [reply kafka/kafka010/src/test/resources/log4j.properties]: Done.

**Comment by sandeshh** [comment]
Any updates on this?

**Comment by chaithu14** [comment]
@tweise I addressed your comments. If you are fine with the changes, please merge it.

**Comment by sandeshh** [comment]
@chaithu14 @tushargosavi Can we merge this PR?, I have a customer waiting for this fix.

**Comment by tushargosavi** [comment]
@chaithu14 I have tested the changes with following configuration
- old application using kafka
- Just updated the dependency in pom.xml with new malhar and was able to run app without any code changes.
- Just updated dependency in pom to xml to use kafka 10 module and was able to run the application with  kafka 0.10 server.

The PR looks good to me. 

Can you fix the test failures? I will merge the pull request after that, If there are no more objections.

**Comment by chaithu14** [comment]
@tushargosavi Thanks for reviewing and testing the changes.
  I'll investigate the test failures and update it. Looks like the Zookeeper client is not started.

**Comment by tweise** [comment]
The output operator should be in the common package, unless there was a change in the producer API.

**Comment by chaithu14** [comment]
@tweise Created separate JIRA(APEXMALHAR-2496) for output. I will open a separate PR for this. I would suggest, first let's merge this PR because there are lot of file changes happened in this PR. 

@tushargosavi I ran the unit tests in cluster, it's working fine. But, these are failing in travis/jenkins. So, I reverted the re-factoring the unit test. But, still the tests are failing. Looks like there is a Kafka server setup issue in travis and also lot's of noisy in logs. Please help me out, how to debug and fix this issue.

**Comment by tweise** [comment]
@chaithu14 this PR is moving around the output operator as well. Unless it does not work with 0.10, please leave it in the common module.

**Comment by chaithu14** [comment]
@tweise Ok. Will verify and add it in common module.

**Comment by chaithu14** [comment]
@tushargosavi @tweise I moved the abstract Kafka output operator to common-package and fixed the test failures.
Please review and merge.

**Comment by tweise** [comment]
Why is KafkaSinglePortOutputOperator not in common? Also can you please fix the parameterized type warning?

**Comment by tweise** [comment]
Why is it necessary to duplicate the input operator test? They seem nearly identical:
```
diff kafka/kafka09/src/test/java/org/apache/apex/malhar/kafka/KafkaInputOperatorTest.java kafka/kafka010/src/test/java/org/apache/apex/malhar/kafka/KafkaInputOperatorTest.java
394,396c394
<       (hasMultiPartition ? "," + l + TEST_KAFKA_BROKER_PORT[0] : "") +
<       (hasMultiCluster ? ";" + l + TEST_KAFKA_BROKER_PORT[1] : "") +
<       (hasMultiCluster && hasMultiPartition ? "," + l + TEST_KAFKA_BROKER_PORT[1] : "");
---
>       (hasMultiCluster ? ";" + l + TEST_KAFKA_BROKER_PORT[1] : "");
```

**Comment by tweise** [comment]
test this please

**Comment by chaithu14** [comment]
@tweise Thanks a lot for reviewing. Yes. KafkaSinglePortOutputOperator can be in kafka-common module, but, not the KafkaSinglePortExactlyOnceOutputOperator. I'll move it.

Fix the parameterized type warning: I'm unable to figure it out. Could you please point it out the code.

Regarding duplicate the input operator test: If i move it to kafka-common module, then it will test only 0.9.* version. I want to test both the versions, that's why, i added it in both modules. If the tests be re-factored, then the tests are failing. I investigated on this, and trying to integrate different embedded servers, but, it's taking time to fix. So, I'll work as a separate task for refactoring the tests.

**Comment by tweise** [comment]
So how about you just keep the test in common and then just extend from it in Kafka010? 

Any idea why the build in Jenkins fails?

**Comment by chaithu14** [comment]
@tweise I tried it, but mostly it's failing due to the creation of Kafka server in travis/jenkins. That's why trying to integrate with below different embedded server's. 
1) EmbeddedKafka(org.apache.apex.malhar.kafka) which is in Kafka module. Limitation is single cluster. Using this, we can't able to test the multi-cluster test cases which are in KafkaInputOperatorTest.
2) KafkaUnit(https://github.com/chbatey/kafka-unit): Limitation - max number of connections allowed from a particular host is 16. Using this, we can test the multi-cluster test cases. But, due to the limitation of connections, can't able to ran all the tests.

It's taking time to fix this, so, will take this as a separate task.

**Comment by chaithu14** [comment]
@tweise I don't have any idea about the Jenkins failed. But, the Kafka tests are passed.

**Comment by tweise** [comment]
It is OK to improve the tests later, but I don't understand why the same code in two different locations would work vs the same unit test code executed from two different test cases not?

**Comment by chaithu14** [comment]
@tweise I don't have an idea about that. But, the failure is due to the server setup.
Please let me know, if there are no more comments, I'll squash the commits and update the PR.

**Comment by tweise** [comment]
Do squash the commits anyways.

What is the issue with the server setup? Why does it work with duplicate code in two places vs. same code called from two places?

**Comment by chaithu14** [comment]
@tweise In the cluster, it's working fine. But, failing it in travis. I don't remembered the exact build, but might be failing due to the below error:
ERROR server.KafkaApis error - [KafkaApi-0] error when handling request Name: TopicMetadataRequest; Version: 0; CorrelationId: 10; ClientId: producer-1; Topics: testtopic0

**Comment by chaithu14** [comment]
@tweise I squashed the commits and updated the PR. If there are no more comments, please merge it.

**Comment by tweise** [comment]
Please address test code in follow-up PR. Output test should probably also be in common, input test code should be de-duped. We need solid test coverage and reliable CI.

**Comment by chaithu14** [comment]
Sure Thomas.

**Comment by tweise** [comment]
@chaithu14 can you point me to the JIRA that will address the test related issues? Did you capture the related info from this PR?  

It looks like the Kafka tests are flaky and that impacts the CI, see:

https://travis-ci.org/apache/apex-malhar/builds/249401536?utm_source=github_status&utm_medium=notification

```
java.lang.AssertionError: received END_TUPLES more than expected.
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.apache.apex.malhar.kafka.KafkaInputOperatorTest$CollectorModule.endWindow(KafkaInputOperatorTest.java:232)
	at com.datatorrent.stram.engine.GenericNode.processEndWindow(GenericNode.java:153)
	at com.datatorrent.stram.engine.GenericNode.run(GenericNode.java:397)
	at com.datatorrent.stram.engine.StreamingContainer$2.run(StreamingContainer.java:1428)
```

**Comment by chaithu14** [comment]
@tweise I haven't created a new JIRA for test related issues. I'll use APEXMALHAR-2162 for tracking these issues. I'm working on re-factor the unit tests and integration with embedded server.

**Pull request merged**
